### PR TITLE
build(common): split ci and pr build definition

### LINKED ISSUE #3126: build(common): split pipeline definition for `autorest.csharp` and `AutoRest C# - Publish`
## Problem

Pipelines `autorest.csharp` and `AutoRest C# - Publish` share the same `eng\pipelines\build.xml`, which cause excessive job run triggered in `auorest.csharp`.

See [this job run](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2191194) in `autorest.csharp`. It's the same as [the one](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2191192) in `AutoRest C# - Publish`. We should try to avoid that.

The root cause is that `autorest.csharp` is supposed to run only for PR, and `AutoRest C# - Publish` is supposed to run only for commits to main branch (e.g. `feature/v3`). So when they share the same build definition YAML file, ``autorest.csharp` is also triggered when there is commit to main branch.

### PR DESCRIPTION
**Pull request by archerzz**
- PR build (e.g. `autorest.csharp` pipeline): still uses `build.xml` which now:
  - only contains build and test steps
  - only triggered by PR, e.g. disable CI build
- CI build (e.g. `AutoRest C# Publish` pipeline): now changes to use `publish.yml` which:
  - only triggered by commits on `feature/v3` branch
  - in the builds we'll publish new versions of `autorest.csharp` and update `azure-sdk-for-net` repo
- extract build and test steps into a template `build-test.yml` which is used by `build.xml` and `publish.yml`

resolve #3126 

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by archerzz** [comment]
For the CI build of `Autorest C# publish`, see the playground run: https://dev.azure.com/azure-sdk/playground/_build/results?buildId=2193980&view=results

**Comment by mikeharder** [comment]
@archerzz, @m-nash: Reading issue #3126, it sounds like what you are trying to do is disable pipeline "autorest.csharp" from running CI builds (when there are new commits to the `feature/v3` branch).  If this is all you want to do, you don't need to change any pipeline YML, you can simply disable the CI trigger in the "autorest.csharp" pipeline settings in DevOps.

But if you think it's better to refactor your pipeline YML in this way, it's also fine with me.

**Comment by m-nash** [comment]
> @archerzz, @m-nash: Reading issue #3126, it sounds like what you are trying to do is disable pipeline "autorest.csharp" from running CI builds (when there are new commits to the `feature/v3` branch). If this is all you want to do, you don't need to change any pipeline YML, you can simply disable the CI trigger in the "autorest.csharp" pipeline settings in DevOps.
> 
> But if you think it's better to refactor your pipeline YML in this way, it's also fine with me.

I do like that this makes two clean pipelines vs having the if/else statements which makes it easy to affect both when you really only wanted to affect one.

@archerzz what are your thoughts on the comment above?

**Comment by archerzz** [comment]
@mikeharder @m-nash Maybe I don't have enough privilege. But I could not find the setting to disable CI trigger, as described in [that stackoverflow Q&A](https://stackoverflow.com/questions/65249158/azure-devops-how-to-disable-ci-trigger-on-a-yaml-template-pipeline). I can only disable the pipeline. That's why I choose to disable it from YAML definition.

Since that also brings cleaner build YAML contents, let's move on this way.

**Review by m-nash — APPROVED** [review]

**Comment by mikeharder** [comment]
> @mikeharder @m-nash Maybe I don't have enough privilege. But I could not find the setting to disable CI trigger, as described in [that stackoverflow Q&A](https://stackoverflow.com/questions/65249158/azure-devops-how-to-disable-ci-trigger-on-a-yaml-template-pipeline). I can only disable the pipeline. That's why I choose to disable it from YAML definition.
> 
> Since that also brings cleaner build YAML contents, let's move on this way.

@weshaggard: Any thoughts on this?  Refactoring the YAML in this way seems fine to me, but EngSys could also disable the CI trigger on this pipeline, or we could maybe grant the necessary permissions to @archerzz.

File: eng/pipelines/publish.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+trigger:
+  branches:
+    include:
+    - feature/*
+
+variables:
+  BlobFeedUrl: https://azuresdkartifacts.blob.core.windows.net/azure-sdk-tools/index.json
```
Comment (weshaggard) [inline eng/pipelines/publish.yml:7]: Why is this needed? This is an old feed which is no longer used.

File: eng/pipelines/publish.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+trigger:
+  branches:
+    include:
+    - feature/*
```
Comment (weshaggard) [inline eng/pipelines/publish.yml:4]: should this only be feature/3?

File: eng/pipelines/build.yml [hunk 3]
```diff
@@ -1,8 +1,4 @@
-trigger:
-  branches:
-    include:
-    - feature/*
-
+trigger: none
 pr:
   branches:
     include:
```
Comment (weshaggard) [inline eng/pipelines/build.yml:4]: BlobFeedUrl property is not needed.

File: eng/pipelines/build-test.yml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+          - script: |
+              npm run test --prefix src/CADL.Extension/Emitter.Csharp
+            displayName: 'Unit Test'
+            workingDirectory: $(Build.SourcesDirectory)/autorest.csharp
+          - task: PublishBuildArtifacts@1
+            condition: succeededOrFailed()
+            displayName: 'Publish Artifacts'
+            inputs:
+              ArtifactName: packages
+          - ${{if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal'))}}:
```
Comment (weshaggard) [inline eng/pipelines/build-test.yml:81]: What is this publish doing? Should it also be moved to the publish.yml?

File: eng/pipelines/build.yml [hunk 5]
```diff
@@ -20,177 +16,6 @@ resources:
       name: Azure/azure-sdk-tools
```
Comment (weshaggard) [inline eng/pipelines/build.yml:16]: This repository reference doesn't look to be used in the build steps anywhere and should be removed.

File: eng/pipelines/build-test.yml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  - stage: 'Build_and_Test'
+    jobs:
+      - job: Build
+        timeoutInMinutes: 120
+        pool:
+          name: azsdk-pool-mms-win-2022-general
+          vmImage: MMS2022Compliant
+        steps:
+          - checkout: self
+          - checkout: azure-sdk-tools
```
Comment (weshaggard) [inline eng/pipelines/build-test.yml:11]: Do we need to checkout this repository any longer? I wouldn't expect so.

File: eng/pipelines/publish.yml (hunk 1 above)
Comment (weshaggard) [reply eng/pipelines/publish.yml:7]: Doesn't look like this variable is used any longer in our template https://dev.azure.com/azure-sdk/internal/_git/azure-sdk-build-tools?path=/pipelines/stages/net-release-to-feed.yml so it should be safe to remove here and in the build.yml file as well.

**Comment by weshaggard** [comment]
I don't mind the split up to help organize things but your split doesn't end up changing the amount of things that run in PR or in CI because the publish steps are guarded by conditions that make them not run during PR. 

As for the split as I mentioned I'm fine with but I did leave some comments for things to clean-up.

File: eng/pipelines/publish.yml (hunk 1 above)
Comment (archerzz) [reply eng/pipelines/publish.yml:7]: Will delete it.

**Comment by archerzz** [comment]
> I don't mind the split up to help organize things but your split doesn't end up changing the amount of things that run in PR or in CI because the publish steps are guarded by conditions that make them not run during PR.
> 
> As for the split as I mentioned I'm fine with but I did leave some comments for things to clean-up.

@weshaggard Actually the main purpose of this PR is to prevent duplicated job run. In [`autorest.csharp` pipeline](https://dev.azure.com/azure-sdk/internal/_build?definitionId=4974), each merge to `feature/v3` will trigger the same job as that in [`Autorest C# Publish` pipeline](https://dev.azure.com/azure-sdk/internal/_build?definitionId=1375&_a=summary). The root cause is in the original YAML file, both CI trigger and PR trigger are defined. That's why I want to split them apart.
image

File: eng/pipelines/build-test.yml (hunk 4 above)
Comment (archerzz) [reply eng/pipelines/build-test.yml:81]: In the `PublishRelease.ps1` will publish artifacts to github/npm, and it will also publish test reports. The logic is tightly coupled. I've created a follow-up issue to move publish artifacts steps to publish: https://github.com/Azure/autorest.csharp/issues/3169

File: eng/pipelines/build.yml (hunk 5 above)
Comment (archerzz) [reply eng/pipelines/build.yml:16]: removed

File: eng/pipelines/build-test.yml (hunk 6 above)
Comment (archerzz) [reply eng/pipelines/build-test.yml:11]: removed

**Comment by weshaggard** [comment]
@archerzz if that is the main reason you don't need to split the yml files to fix that. You can simply edit on of the pipelines to disable the CI triggers in the DevOps UI. So that might be an easier option if we don't want split these.

**Comment by archerzz** [comment]
> @archerzz if that is the main reason you don't need to split the yml files to fix that. You can simply edit on of the pipelines to disable the CI triggers in the DevOps UI. So that might be an easier option if we don't want split these.

Thanks. I found now I can edit the pipeline triggers. I've disabled the CI trigger for `autorest.csharp` pipeline. So this PR is unnecessary. Closing.
